src/NuGet.Core/NuGet.Configuration/PackageNamespaces/SearchTree.cs

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace NuGet.Configuration
+{
+    internal class SearchTree
+    {
+        private readonly SearchNode _root;
+
+        internal SearchTree(PackageNamespacesConfiguration configuration)
+        {
+            if (configuration == null)
+            {
+                throw new ArgumentNullException(nameof(configuration));
+            }
+
+            _root = new SearchNode();
+
+            foreach (string packageSourceKey in configuration.Namespaces.Keys)
+            {
+                foreach (string nugetNamespace in configuration.Namespaces[packageSourceKey])
+                {
+                    Add(packageSourceKey, nugetNamespace);
```

Comment:
The key just retrieves the source uri doesn't it? Perhaps we use the most portable value such as the source uri as kartheek proposes?

Reply:
@nkolev92 @JonDouglas 
To make this work we need to make more changes in `PackageNamespacesConfiguration`. It need to know about `source Uri` to `packageSource key` mapping, currently it doesn't. I'm not sure if it's desired thing.

Reply:
I still doubt if we need `source uri`. Because `nuget.config` should have knowledge of passing `sourceUri` values so it can match in package name spaces section. Then it defeats the purpose of `-Source` parameter.

Reply:
@nkolev92 
I love to hear your opinion here.

Reply:
@kartheekp-ms @nkolev92 
I agree we need to break into smaller tasks . I created follow up issue for this since rebasing and resolving merge conflict becoming more harder as time passes . https://github.com/NuGet/Home/issues/10948